### PR TITLE
Refs #16468 - Play ansible roles through smart proxy

### PR DESCRIPTION
**Pull request by iNecas**
Opening a separate PR to unblock testing + additional work on this.

I've separated the working dir from ansible dir, as we don't need to have
access to the ansible dir (where the roles are), while we need some space
for writing inventory files.

Similar to https://github.com/theforeman/foreman_ansible/pull/40, I would like
to:
  - [x] - extract the common code with smart_proxy_remote_execution_ssh and smart_proxy_dynflow,
and also make the core parts usable
  - [x] - make the core working both with foreman and smart proxy, making it easier to run
  simpler setups.

### CODE REVIEW

**Comment by dLobatog** [comment]
I'm utterly confused about this PR. We can already use 99% of the existing code to run roles, this PR does a lot more (creates an inventory (without parameters), creates the playbook, etc...). That's unnecessary with foreman_ansible who can just provide the playbook & inventory, the proxy just needs to be smart enough to run it & provide output, nothing else.

**Comment by iNecas** [comment]
The motivation was to extract the playbook run mechanism from the foreman so that one can achieve the same without even needing to do so in Foreman, or even smart proxy. This brings more flexibility on the ways ones performing the Foreman's playbook execution. Being able to drive the playbooks execution outside of Foreman allows:

  1. more options when dealing with scaling
  2. ability to drive it by different mechanisms: cron, Tower, Jenkins

I'm ok with extracting the common code into separate gem, to use the same logic both on foreman server and smart proxy,and keeping just the proxying mechanism in proxy.

**Comment by iNecas** [comment]
I've moved all code that I could to formen_ansible, what remained was just configuration of settings.

File: bundler.plugins.d/remote_execution_ansible_core.rb [hunk 1]
```diff
@@ -0,0 +1 @@
+gem 'smart_proxy_ansible_core'
```
Comment (dLobatog) [inline bundler.plugins.d/remote_execution_ansible_core.rb]: Wrong name?

Comment (iNecas) [inline bundler.plugins.d/remote_execution_ansible_core.rb]: updated

File: bundler.plugins.d/foreman_ansible_core.rb [hunk 2]
```diff
@@ -0,0 +1 @@
+gem 'foreman_ansible_core'
```
Comment (dLobatog) [inline bundler.plugins.d/foreman_ansible_core.rb]: This ought to be a requirement in the  gemspec too, I think, the plugin doesn't work without it

File: lib/smart_proxy_ansible/plugin.rb [hunk 3]
```diff
@@ -4,7 +4,17 @@ class Plugin < Proxy::Plugin
...
-    default_settings :ansible_working_dir => '~/.foreman-ansible'
     plugin :ansible, Proxy::Ansible::VERSION
+
+    after_activation do
+      begin
+        require 'smart_proxy_dynflow_core'
+        require 'foreman_ansible_core'
+        ForemanAnsibleCore.initialize_settings(Proxy::Ansible::Plugin.settings.to_h)
+      rescue LoadError => _
+        # Dynflow core is not available in the proxy, will be handled
```
Comment (dLobatog) [inline lib/smart_proxy_ansible/plugin.rb]: Nothing shows up when I put binding.pry here, I don't think the problem has to do with loading dynflow

Comment (iNecas) [inline lib/smart_proxy_ansible/plugin.rb]: The problem might be with `ForemanAnsibleCore.initialize_settings(Proxy::Ansible::Plugin.settings.to_h)` not getting called or failing… I have tried a fresh proxy setup with this https://gist.github.com/iNecas/dcc21b1eccce751b75baacf8e21b1b25 and was still not able to reproduce. I tried also ruby-2.3.0-preview1, still worked fine. Maybe something in logs?

Comment (dLobatog) [inline lib/smart_proxy_ansible/plugin.rb]: I was under the impression that  :branch => 'integrate-foreman-tasks-core' was not necessary for smart_proxy_dynflow - now it works, thanks

File: bundler.plugins.d/foreman_ansible_core.rb (hunk 2 above)
Comment (dLobatog) [inline bundler.plugins.d/foreman_ansible_core.rb]: Sorry, that confused me a bit. How do you enforce that `foreman_ansible_core` is loaded with this plugin (or `smart_proxy_dynflow`)? Currently from what I see it's up to the user to set this line in bundler.plugins.d in the proxy.

**Comment by dLobatog** [comment]
Ready to merge IMO, pending two things:

* The issue tracker number should be different for different projects
* I'd like to understand how the requirement of `foreman_ansible_core` is enforced.

Thanks @iNecas , huge effort here.

**Comment by iNecas** [comment]
I have squashed the commits and used separate issue for the smart proxy changes.

The dependency on foreman_ansible_core will get in by packaging or installer. The reason is the rhel6, where we can't load it from the smart proxy, and need separate process for this. On non-rhel systems, we put the dependency in packaging (we do that in smart proxy rex). If we had something else in the foreman_ansible_core except dynflow-related stuff, we could do that, and we can do that once that happens (probably will when extracting roles listing etc), but for now, I would keep it separate.

**Comment by dLobatog** [comment]
Thank you @iNecas and @adamruzicka !

**Pull request merged**